src/_zkapauthorizer/_storage_server.py

```diff
@@ -695,3 +701,100 @@ def get_stat(sharepath):
             return stat_slot
         else:
             return stat_bucket
+
+
+def add_leases_for_writev(storage_server, storage_index, secrets, tw_vectors, now):
+    """
+    Add a new lease using the given secrets to all shares written by
+    ``tw_vectors``.
+    """
+    for (sharenum, sharepath) in get_all_share_paths(storage_server, storage_index):
+        testv, datav, new_length = tw_vectors.get(sharenum, (None, b"", None))
+        if datav or (new_length is not None):
+            # It has data or a new length - it is a write.
+            if share_has_active_leases(storage_server, storage_index, sharenum, now):
+                # It's fine, leave it be.
+                continue
+
+            # Aha.  It has no lease that hasn't expired.  Give it one.
+            (write_enabler, renew_secret, cancel_secret) = secrets
+            share = get_share_file(sharepath)
+            share.add_or_renew_lease(
+                LeaseInfo(
+                    owner_num=1,
+                    renew_secret=renew_secret,
+                    cancel_secret=cancel_secret,
+                    expiration_time=now
+                    + ZKAPAuthorizerStorageServer.LEASE_PERIOD.total_seconds(),
+                    nodeid=storage_server.my_nodeid,
+                ),
+            )
+
+
+def get_share_path(storage_server, storage_index, sharenum):
+    # type: (StorageServer, bytes, int) -> FilePath
+    """
+    Get the path to the given storage server's storage for the given share.
+    """
+    return (
+        FilePath(storage_server.sharedir)
+        .preauthChild(storage_index_to_dir(storage_index))
+        .child(u"{}".format(sharenum))
+    )
+
+
+def share_has_active_leases(storage_server, storage_index, sharenum, now):
+    # type: (StorageServer, bytes, int, float) -> bool
+    """
+    Determine whether the given share on the given server has an unexpired
+    lease or not.
+
+    :return: ``True`` if it has at least one unexpired lease, ``False``
+        otherwise.
+    """
+    sharepath = get_share_path(storage_server, storage_index, sharenum)
+    share = get_share_file(sharepath.path)
+    return any(lease.get_expiration_time() > now for lease in share.get_leases())
+
+
+def get_writev_price(storage_server, pass_value, storage_index, tw_vectors, now):
+    # type: (StorageServer, int, bytes, TestWriteVectors, float) -> int
+    """
+    Determine the price to execute the given test/write vectors.
+    """
+    # Find the current size of shares being written.
+    current_sizes = dict(
+        get_share_sizes(
+            storage_server,
+            storage_index,
+            # Consider the size of *all* shares even if they're not being
+            # written.  If they have an unexpired lease then we can apply some
+            # or all of the remainder of the value of that lease towards this
+            # operation.
```

Comment:
I'm not sure how meaningful this comment is. I was going to say that sharefiles that aren't touched are going to have the same size before and after in `get_required_new_passes_for_mutable_write`, but I guess that size is taken into account when `required_passes` is rounding the number of passes required. However, I think that can only ever change the number of passes required by ±1, but that doesn't really seem like applying the remaining value of those leases to anything we charge here.

---

As an aside, it does feel a little weird that existing share can have that effect. I guess a client can always zero out the size of any shares it doesn't care about.

Reply:
I expanded on this comment in 6789754d and I think I gave an example where the required number of passes can change by more than 1... but at a higher level, I agree this is weird behavior.  It's also a bit of a tangent to the issue this PR is meant to address.  I think that what happened is when I made the Hypothesis-based test for mutables more comprehensive, it uncovered another existing bug where the client and server disagreed on the price calculation.  The client was already counting all existing shares but the server was not.

It probably would have been better to split this out into a separate ticket/PR but I forgot to.  I still could, but maybe since you already spent the time reading the diff it's not particularly worthwhile now.

I think the behavior implemented here is not ideal but it is better than before (because it will not cause spending errors due to client/server disagreement in this case anymore).

Reply:
> I don't think your example works as you describe, since we don't ever take the sum of quantized values.

Okay.  I think you're right.  Eventually I decided to change the logic to go the other way.  Now the client and server both agree to ignore existing shares.